interfaces/builtin/gpio-control.go

```diff
@@ -0,0 +1,49 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016-2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+const gpioControlSummary = `allows control of all GPIOs`
+
+const gpioControlBaseDeclarationSlots = `
+  gpio-control:
+    allow-installation:
+      slot-snap-type:
+        - core
```

Comment:
I am not a GPIO expert, but because this interface grants export/unexport and GPIOs are "General Purpose Input/Output" and can be used with all kinds of disparate hardware (or even to power off machines, work with NAND and other functionality, see aforementioned kernel docs), I'm uneasy with this being a 'normal' interface. It grants more than one might expect and on some devices, depending on how they are wired, something along the lines of device ownership.

Furthermore (cc @pedronis), on a device with a gadget snap that assigns GPIOs to specific snaps, a snap with this interface would be able to wreck havoc for these other snaps by, at a minimum, unexporting the hardware. ISTR discussions where we really didn't want snaps to have this kind of access, but I forget the details (the conversations didn't seem to have happened in github PRs :/); I suspect it may have had to do with this. That said, it very well may have been a situation where we didn't want control like this to be the default proposition for snaps, but now that we have the gpio interface for normal use (and hotplug available as use cases dictate), adding a specialty interface like this probably makes sense.

In all, I think this needs an installation constraint and we should treat the interface as superprivileged. Please adjust this to be:
```
// Controlling all aspects of GPIO pins can potentially impact other snaps and
// grant wide access to specific hardware and the system, so treat as
// super-privileged
const gpioControlBaseDeclarationSlots = `
  gpio-control:
    allow-installation: false
    deny-auto-connection: true
`
```

Reply:
I agree this should be super-privileged interface. It indeed has inherited risks and can cause serious problems if misused. gpio slots for each particular pin exposed by gadget snaps are still go-to solution for us. This interface should address situations when our standard measures fall short and there is no easy way to simply alter behaviour of used software stack ( as in our example it's zigbee stack)
Changing values accordingly

Reply:
this change doesn't seem to have happened yet

Reply:
Seems there is some confusion in that snippet by @jstrand, what we need also is a ` gpioControlBaseDeclarationPlugs`

Reply:
Sorry, yes, I typoed the snippet. This is what I mean:
```
// Controlling all aspects of GPIO pins can potentially impact other snaps and
// grant wide access to specific hardware and the system, so treat as
// super-privileged
const gpioControlBaseDeclarationPlugs = `
  gpio-control:
    allow-installation: false
    deny-auto-connection: true
```

Reply:
@jdstrand I have updated code not to include following:
```
// Controlling all aspects of GPIO pins can potentially impact other snaps and
// grant wide access to specific hardware and the system, so treat as
// super-privileged
const gpioControlBaseDeclarationPlugs = `
  gpio-control:
    allow-installation: false
    deny-auto-connection: true
`

const gpioControlBaseDeclarationSlots = `
  gpio-control:
    allow-installation:
      slot-snap-type:
        - core
    deny-auto-connection: true
`
```

Reply:
@pedronis mind I had wrong file name syntax so file name changed from `interfaces/builtin/gpio-control.go` to `interfaces/builtin/gpio_control.go`